packages/cli/generators/controller/index.js

```diff
@@ -181,6 +181,12 @@ module.exports = class ControllerGenerator extends ArtifactGenerator {
         when: this.artifactInfo.idType === undefined,
         default: 'number',
       },
+      {
+        type: 'confirm',
+        name: 'idRequired',
+        message: 'Is the id required when creating an instance?',
+        default: false,
```

Comment:
Can we check the `generated` flag for the id property? See https://github.com/strongloop/loopback-next/pull/3643

Reply:
@raymondfeng The fix in this PR is in the **controller** generator, so it doesn't have the `generated` flag (#3643 is modifying **model** generator)

Reply:
Can the controller code generator have access the model definition to check?

Reply:
It would be great if `lb4 controller` could automatically infer whether the `id` property is required or forbidden when creating new models. At the same time, I think the current proposal is a meaningful improvement, therefore it should be landed. We can improve the generator to parse `id` property definition later.

IIRC, there are multiple ways how to end up with a model that requires `id` values on creation. The model-level setting `forceId` plays an important role.

```ts
@model({settings: {forceId: false}})
export class TodoWithId extends Entity {
  @property({
    type: 'number',
    id: true,
  })
  id?: number;

  // ...
}
```

When `forceId` is disabled, clients have can either supply their own id, or let the database generate a new unique id for them. (In the schema, the `id` property should be shown as optional.) That's my understanding of how things should work, I did not run a test to verify the actual behavior.

Reply:
@raymondfeng 
> Can the controller code generator have access the model definition to check?

It should have access to the model def if the file exist. Then use `ts-morph` to read the element. 

And I agree with @bajtos that we probably can land this PR first, since in some situations, if `idRequired` cannot be inferred from the model class, a prompt is still needed.

Like the [id name prompt](https://github.com/strongloop/loopback-next/blob/master/packages/cli/generators/controller/index.js#L170), when `id` cannot be inferred from the `artifactInfo`, the prompt will ask for it.

I will submit another PR to read the `idRequired` from model file and add a `when` to the prompt. I believe there would be more discussion about how we infer the flag :)